x-pack/plugins/lens/public/datasources/form_based/operations/definitions/formula/context_variables.tsx

```diff
@@ -0,0 +1,269 @@
+// TODO: split layer_helpers util into pure/non-pure functions to avoid issues with tests
+export function getColumnOrder(layer: FormBasedLayer): string[] {
+  const entries = Object.entries(layer.columns);
+  entries.sort(([idA], [idB]) => {
+    const indexA = layer.columnOrder.indexOf(idA);
+    const indexB = layer.columnOrder.indexOf(idB);
+    if (indexA > -1 && indexB > -1) {
+      return indexA - indexB;
+    } else if (indexA > -1) {
+      return -1;
+    } else {
+      return 1;
+    }
+  });
+
+  const [aggregations, metrics] = partition(entries, ([, col]) => col.isBucketed);
+
+  return aggregations.map(([id]) => id).concat(metrics.map(([id]) => id));
+}
+
+// Copied over from helpers
+export function isColumnOfType<C extends GenericIndexPatternColumn>(
+  type: C['operationType'],
+  column: GenericIndexPatternColumn
+): column is C {
+  return column.operationType === type;
+}
+
+export interface ContextValues {
+  dateRange?: DateRange;
+  now?: Date;
+  targetBars?: number;
+}
+
+export interface TimeRangeIndexPatternColumn extends ReferenceBasedIndexPatternColumn {
+  operationType: 'time_range';
+}
+
+function getTimeRangeFromContext({ dateRange }: ContextValues) {
+  return dateRange ? moment(dateRange.toDate).diff(moment(dateRange.fromDate)) : 0;
+}
+
+function getTimeRangeErrorMessages(
+  layer: FormBasedLayer,
+  columnId: string,
+  indexPattern: IndexPattern,
+  dateRange?: DateRange | undefined
+) {
+  const errors = [];
+  if (!indexPattern.timeFieldName) {
+    errors.push(
+      i18n.translate('xpack.lens.indexPattern.dateRange.dataViewNoTimeBased', {
+        defaultMessage: 'The current dataView is not time based',
+      })
+    );
+  }
+  if (!dateRange) {
+    errors.push(
+      i18n.translate('xpack.lens.indexPattern.dateRange.noTimeRange', {
+        defaultMessage: 'The current time range interval is not available',
+      })
+    );
+  }
+  return errors.length ? errors : undefined;
+}
+
+export const timeRangeOperation = createContextValueBasedOperation<TimeRangeIndexPatternColumn>({
+  type: 'time_range',
+  label: 'Time range',
+  description: i18n.translate('xpack.lens.indexPattern.timeRange.documentation.markdown', {
+    defaultMessage: `
+The current time range expressed in milliseconds (ms).
+    `,
+  }),
+  getContextValue: getTimeRangeFromContext,
+  getErrorMessage: getTimeRangeErrorMessages,
+});
+
+export interface NowIndexPatternColumn extends ReferenceBasedIndexPatternColumn {
+  operationType: 'now';
+}
+
+function getNowFromContext({ now }: ContextValues) {
+  return now == null ? Date.now() : +now;
+}
+function getNowErrorMessage() {
+  return undefined;
+}
+
+export const nowOperation = createContextValueBasedOperation<NowIndexPatternColumn>({
+  type: 'now',
+  label: 'Current now',
+  description: i18n.translate('xpack.lens.indexPattern.now.documentation.markdown', {
+    defaultMessage: `
+  The current now moment used in Kibana expressed in milliseconds (ms).
```

Comment:
I'm unsure what `now moment` means?

Reply:
I meant the instant value of `now`, expressed in `ms`